### PR TITLE
Generate date localizations

### PR DESCRIPTION
**Pull request by yjbanov**
Add a script that extracts the sub-set of `intl` date data for locales supported by Flutter (currently we pull in _all_ locale data). Extracted data goes into a new file `i18n/date_localizations.dart`.

Here are the numbers for building `complex_layout`:

`aot_snapshot_size_isolate`:

Before #12324: 1832972 bytes
After #12324: 1941640 bytes
After #12466: 1867322 bytes

`aot_snapshot_size_rodata`:

Before #12324: 1758568 bytes
After #12324: 1892776 bytes
After #12466: 1794616 bytes

### CODE REVIEW

File: dev/tools/gen_date_localizations.dart [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:args/args.dart' as args;
+import 'package:path/path.dart' as path;
+
+const String _kCommandName = 'gen_date_localizations.dart';
+
+Future<Null> main(List<String> rawArgs) async {
+  final bool writeToFile = (new args.ArgParser()..addFlag('write', abbr: 'w', defaultsTo: false)).parse(rawArgs)['write'];
```
Comment (HansMuller) [inline dev/tools/gen_date_localizations.dart]: It would be good to update gen_localizations.dart to be consistent vis this 'w' flag.

Give the 'w' flag a --overwrite synonym (like dartfmt)?

Likewise for the requirement (enforced below) that the app be run from the isRepoRoot.

File: dev/tools/gen_date_localizations.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  final bool dotPackagesExists = dotPackagesFile.existsSync();
+
+  if (!dotPackagesExists) {
+    _fatal(
+      'File not found: ${dotPackagesFile.path}. $_kCommandName must be run '
+      'after a successful "flutter update-packages".'
+    );
+  }
+
+  final String pathToIntl = dotPackagesFile
```
Comment (HansMuller) [inline dev/tools/gen_date_localizations.dart]: This is
.pretty
.slick

File: dev/tools/gen_date_localizations.dart [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    return '$json';
+
+  if (json is String) {
+    if (json.contains("'"))
+      return 'r"""$json"""';
+    else
+      return "r'''$json'''";
+  }
+
+  if (json is Iterable)
```
Comment (HansMuller) [inline dev/tools/gen_date_localizations.dart]: You could generate const <String> or const <int> here with just a little more code.

File: dev/tools/gen_date_localizations.dart [hunk 4]
```diff
@@ -0,0 +1,147 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (HansMuller) [inline dev/tools/gen_date_localizations.dart]: There should be a comment about what this app does, when it needs to be run, how to run it.

File: dev/tools/gen_date_localizations.dart [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+// This file has been automatically generated.  Please do not edit it manually.
+// To regenerate run (omit -w to print to console instead of the file):
+// dart dev/tools/gen_date_localizations.dart -w
+
+'''
+);
+  buffer.writeln('const Map<String, dynamic> dateSymbols = const <String, dynamic> {');
+  symbolFiles.forEach((String locale, File data) {
+    if (materialLocales.contains(locale))
+      buffer.writeln(_jsonToMapEntry(locale, JSON.decode(data.readAsStringSync())));
```
Comment (HansMuller) [inline dev/tools/gen_date_localizations.dart]: It would be nice to declare which fields we actually need and only write those. That would also give us the opportunity to verify that the fields we expect are still defined.

File: packages/flutter/lib/src/material/material_localizations.dart [hunk 6]
```diff
@@ -8,8 +8,11 @@ import 'package:flutter/foundation.dart';
 import 'package:flutter/widgets.dart';
 import 'package:intl/intl.dart' as intl;
 import 'package:intl/date_symbols.dart' as intl;
-import 'package:intl/date_symbol_data_local.dart' as intl_local_date_data;
+// TODO: remove internal import when https://github.com/dart-lang/intl/issues/145 is fixed.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/material_localizations.dart]: TODO(yegor):

**Review by HansMuller — COMMENTED** [review]
It would be good to hold off on trying to land this until https://github.com/flutter/flutter/pull/12410 has landed since the updates to material_localizations.dart will end up in the flutter_localizations package

File: .idea/modules.xml [hunk 7]
```diff
@@ -10,6 +10,7 @@
       <module fileurl="file://$PROJECT_DIR$/packages/flutter_test/flutter_test.iml" filepath="$PROJECT_DIR$/packages/flutter_test/flutter_test.iml" />
       <module fileurl="file://$PROJECT_DIR$/packages/flutter_tools/flutter_tools.iml" filepath="$PROJECT_DIR$/packages/flutter_tools/flutter_tools.iml" />
       <module fileurl="file://$PROJECT_DIR$/examples/hello_world/hello_world.iml" filepath="$PROJECT_DIR$/examples/hello_world/hello_world.iml" />
+      <module fileurl="file://$PROJECT_DIR$/dev/tools/tools.iml" filepath="$PROJECT_DIR$/dev/tools/tools.iml" />
```
Comment (Hixie) [inline .idea/modules.xml]: this presumably doesn't belong in this PR?

File: packages/flutter/lib/src/material/i18n/date_localizations.dart [hunk 8]
```diff
@@ -0,0 +1,2936 @@
...
+    'WEEKENDRANGE': const <dynamic>[5, 6],
+    'FIRSTWEEKCUTOFFDAY': 5,
+    'DATETIMEFORMATS': const <dynamic>[
+      r'''{1} {0}''',
+      r'''{1} {0}''',
+      r'''{1} {0}''',
+      r'''{1} {0}'''
+    ],
+  },
+};
```
Comment (Hixie) [inline packages/flutter/lib/src/material/i18n/date_localizations.dart]: Why maps instead of just hard-coded classes? It would probably be a lot more efficient that way.

**Comment by HansMuller** [comment]
LGTM

File: packages/flutter/lib/src/material/i18n/date_localizations.dart (hunk 8 above)
Comment (yjbanov) [reply packages/flutter/lib/src/material/i18n/date_localizations.dart]: We are constructing the [DateSymbols class](https://www.dartdocs.org/documentation/intl/0.15.2/date_symbols/DateSymbols-class.html) from this data once during initialization. It's objects from then on.

We could reduce initialization time if we could create `const` objects. Unfortunately, `DateSymbols` is not `const`, so it has to be initialized at runtime. I can look into that next, but it doesn't look doable as part of this PR.
